File: 4_Sum.cpp
class Solution {
public:
    /**
     * @param numbers: Give an array numbersbers of n integer
     * @param target: you need to find four elements that's sum of target
     * @return: Find all unique quadruplets in the array which gives the sum of 
     *          zero.
     */
    vector<vector<int> > fourSum(vector<int> nums, int target) {
        // write your code here
        unordered_map<int, vector<pair<int, int>>> table;
        sort(nums.begin(), nums.end());
        int length = nums.size();
        for (int i = 0; i < length; ++i){
            for (int j = i + 1; j < length; ++j){
                bool is_duplicate = false;
                for (auto& p : table[nums[i] + nums[j]]){
                    if (p.first == i){
                        is_duplicate = true;
                        break;
                    }
                }
                if (!is_duplicate){
                    table[nums[i] + nums[j]].emplace_back(pair<int,int>(i, j));
                }
            }
        }
        unordered_set<string> check_unique;
        vector<vector<int>> ret;
        for (int i = 0; i < length; ++i){
            for (int j = i + 1; j < length; ++j){
                int diff = target - nums[i] - nums[j];
                if (table.find(diff) != table.end()){
                    for (auto& p : table[diff]){
                        if (j < p.first){
                            vector<int> tmp = {nums[i], nums[j], nums[p.first], nums[p.second]};
                            if (check_unique.emplace(vector_to_string(tmp)).second){
                                ret.emplace_back(tmp);
                            }
                        }
                    }
                }
            }
        }
        
        return ret;
    }
    string vector_to_string(vector<int> v){
        string ret;
        for (auto& i : v){
            ret += i;
            ret += ' ';
        }
        return ret;
    }
};

File: 4Sum.py
class Solution:
    """
    @param numbersbers : Give an array numbersbersbers of n integer
    @param target : you need to find four elements that's sum of target
    @return : Find all unique quadruplets in the array which gives the sum of 
              zero.
    """
    def fourSum(self ,numbers, target):
        # write your code here
        if not numbers:
            return []
        # numbers.sort()
        # length = len(numbers)
        # ans = []
        # for i in range(length - 3):
        #     if i > 0 and numbers[i] == numbers[i - 1]:
        #         continue
        #     for j in range(i + 1, length - 2):
        #         if j > i + 1 and numbers[j] == numbers[j - 1]:
        #             continue
        #         left, right = j + 1, length - 1
        #         while left < right:
        #             total = numbers[i] + numbers[j] + numbers[left] + numbers[right]
        #             if total > target:
        #                 right -= 1
        #                 # while left < right and numbers[right] == numbers[right + 1]:
        #                 #     right -= 1
        #             elif total < target:
        #                 left += 1
        #                 # while left < right and numbers[left] == numbers[left - 1]:
        #                 #     left += 1
        #             else:
        #                 ans.append([numbers[i], numbers[j], numbers[left], numbers[right]])
        #                 left += 1
        #                 right -= 1
        #                 while left < right and numbers[left] == numbers[left - 1]:
        #                     left += 1
        #                 while left < right and numbers[right] == numbers[right + 1]:
        #                     right -= 1
        # return ans
        numbers.sort()
        length = len(numbers)
        hashVal = collections.defaultdict(set)
        for i in range(length):
            for j in range(i + 1, length):
                hashVal[numbers[i] + numbers[j]].add((i, j))
        ans = set()
        for i in range(length):
            if i > 0 and numbers[i] == numbers[i - 1]:
                continue
            for j in range(i + 1, length):
                if j > i + 1 and numbers[j] == numbers[j - 1]:
                    continue
                total = numbers[i] + numbers[j]
                if target - total in hashVal:
                    indexs = hashVal[target - total]
                    for index in indexs:
                        if j < index[0]:
                            ans.add((numbers[i], numbers[j], numbers[index[0]], numbers[index[1]]))
        return [list(v) for v in ans]
                
                
